core/src/main/scala/kafka/log/Log.scala

```diff
@@ -421,25 +421,41 @@ class Log(@volatile var dir: File,
 
   private def loadProducerState(lastOffset: Long): Unit = lock synchronized {
     info(s"Loading producer state from offset $lastOffset for partition $topicPartition")
-    val currentTimeMs = time.milliseconds
-    producerStateManager.truncateAndReload(logStartOffset, lastOffset, currentTimeMs)
-
-    // only do the potentially expensive reloading of the last snapshot offset is lower than the
-    // log end offset (which would be the case on first startup) and there are active producers.
-    // if there are no active producers, then truncating shouldn't change that fact (although it
-    // could cause a producerId to expire earlier than expected), so we can skip the loading.
-    // This is an optimization for users which are not yet using idempotent/transactional features yet.
-    if (lastOffset > producerStateManager.mapEndOffset || !producerStateManager.isEmpty) {
-      logSegments(producerStateManager.mapEndOffset, lastOffset).foreach { segment =>
-        val startOffset = math.max(segment.baseOffset, producerStateManager.mapEndOffset)
-        val fetchDataInfo = segment.read(startOffset, Some(lastOffset), Int.MaxValue)
-        if (fetchDataInfo != null)
-          loadProducersFromLog(producerStateManager, fetchDataInfo.records)
+
+    if (producerStateManager.latestSnapshotOffset.isEmpty) {
+      // if there are no snapshots to load producer state from, we assume that the brokers are
+      // being upgraded, which means there would be no previous idempotent/transactional producers
+      // to load state for. To avoid an expensive scan through all of the segments, we take
+      // empty snapshots from the start of the last two segments and the last offset. The purpose
+      // of taking the segment snapshots is to avoid the full scan in the case that the log needs
+      // truncation.
+      val nextLatestSegmentBaseOffset = Option(segments.lowerEntry(activeSegment.baseOffset)).map(_.getValue.baseOffset)
+      val offsetsToSnapshot = Seq(nextLatestSegmentBaseOffset, Some(activeSegment.baseOffset), Some(lastOffset))
+      offsetsToSnapshot.flatten.distinct.foreach { offset =>
```

Comment:
Not sure if distinct preserver ordering. We probably want to sort the distinct offsets in ascending order so that updateMapEndOffset() is called in offset order?